contracts/adapters/OptimisticAuctionRebalanceExtension.sol

```diff
@@ -0,0 +1,390 @@
+/*
+    Copyright 2023 Index Coop
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    SPDX-License-Identifier: Apache License, Version 2.0
+*/
+
+pragma solidity 0.6.10;
+pragma experimental "ABIEncoderV2";
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from  "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+
+import { AddressArrayUtils } from "../lib/AddressArrayUtils.sol";
+
+import { IAuctionRebalanceModuleV1 } from "../interfaces/IAuctionRebalanceModuleV1.sol";
+import { IBaseManager } from "../interfaces/IBaseManager.sol";
+import { ISetToken } from "../interfaces/ISetToken.sol";
+import {AuctionRebalanceExtension} from "./AuctionRebalanceExtension.sol";
+import {AncillaryData } from "../lib/AncillaryData.sol";
+import { AssetAllowList } from "../lib/AssetAllowList.sol";
+import {OptimisticOracleV3Interface} from "../interfaces/OptimisticOracleV3Interface.sol";
+
+/**
+ * @title BaseOptimisticAuctionRebalanceExtension
+ * @author Index Coop
+ *
+ * @dev The contract extends `BaseAuctionRebalanceExtension` by adding an optimistic oracle mechanism for validating rules on the proposing and executing of rebalances. 
+ * It allows setting product-specific parameters for optimistic rebalancing and includes callback functions for resolved or disputed assertions.
+ */
+contract OptimisticAuctionRebalanceExtension is  AuctionRebalanceExtension, AssetAllowList {
+    using AddressArrayUtils for address[];
+    using SafeERC20 for IERC20;
+
+    /* ============ Events ============ */
+
+    event ProductSettingsUpdated(
+        IERC20 indexed setToken,
+        address indexed manager,
+        OptimisticRebalanceParams optimisticParams,
+        bytes32 indexed rulesHash
+    );
+    event RebalanceProposed(
+        ISetToken indexed setToken,
+        IERC20 indexed quoteAsset,
+        address[] oldComponents,
+        address[] newComponents,
+        AuctionExecutionParams[] newComponentsAuctionParams,
+        AuctionExecutionParams[] oldComponentsAuctionParams,
+        bool shouldLockSetToken,
+        uint256 rebalanceDuration,
+        uint256 positionMultiplier
+    );
+
+    event AssertedClaim(
+        IERC20 indexed setToken,
+        address indexed _assertedBy,
+        bytes32 indexed rulesHash,
+        bytes32 _assertionId,
+        bytes _claimData
+    );
+
+    event ProposalDeleted(
+        bytes32 assertionID, 
+        Proposal indexed proposal
+        );
```

Comment:
may i request we lint this contract to make spacing consistent?

Reply:
Manually fixed the indentation in this case. Any automatic linter you can recommend for solidity ?

Reply:
addressing in #157